File: frontend/src/types/runway.ts
/**
 * Runway-related type definitions
 * 跑道参数相关类型定义
 */

/**
 * Coordinate representation
 * 坐标点表示
 */
export interface Coordinate {
  /** Latitude in degrees (纬度) */
  latitude: number;
  /** Longitude in degrees (经度) */
  longitude: number;
}

/**
 * Coordinate system type
 * 坐标系类型
 */
export type CoordinateSystem = 'WGS84' | 'GCJ02';

/**
 * Runway parameters for flight circuit calculation
 * 跑道参数定义
 */
export interface RunwayParams {
  /** Runway center point coordinate (跑道中心点坐标) */
  coordinate: Coordinate;
  /** Magnetic bearing in degrees (磁方位角，0-360度) */
  magneticBearing: number;
  /** Runway length in meters (跑道长度，单位：米) */
  length: number;
  /** Runway width in meters (跑道宽度，0或不填=自动根据飞行区指标) */
  runwayWidth?: number;
  /** Runway elevation in meters, 0.1m precision (跑道标高，单位：米，精确到0.1米) */
  elevation: number;
  /** Coordinate system used for input (输入坐标系类型) */
  coordinateSystem: CoordinateSystem;
}

/**
 * Validation error for runway parameters
 * 跑道参数校验错误
 */
export interface RunwayValidationError {
  /** Field name that has error (错误字段名) */
  field: string;
  /** Error message (错误消息) */
  message: string;
  /** Error severity (错误级别) */
  severity: 'error' | 'warning';
}

/**
 * Runway validation result
 * 跑道参数校验结果
 */
export interface RunwayValidationResult {
  /** Whether parameters are valid (是否有效) */
  isValid: boolean;
  /** Validation errors (校验错误列表) */
  errors: RunwayValidationError[];
}

/**
 * Default runway parameters
 * 默认跑道参数
 */
export const DEFAULT_RUNWAY_PARAMS: RunwayParams = {
  coordinate: { latitude: 30.2741, longitude: 120.1551 },
  magneticBearing: 180,
  length: 800,
  elevation: 20,
  coordinateSystem: 'WGS84',
};

/**
 * Coordinate validation constraints
 * 坐标校验约束
 */
export const COORDINATE_CONSTRAINTS = {
  latitude: { min: -90, max: 90 },
  longitude: { min: -180, max: 180 },
};

/**
 * Magnetic bearing constraints
 * 磁方位角约束
 */
export const MAGNETIC_BEARING_CONSTRAINTS = {
  min: 0,
  max: 360,
};

/**
 * Runway length constraints (meters)
 * 跑道长度约束（米）
 */
export const RUNWAY_LENGTH_CONSTRAINTS = {
  min: 200, // 最小200米（轻型飞机）
  max: 5000, // 最大5000米
  recommended_min: 600, // 推荐最小600米（塞斯纳172）
};

/**
 * Elevation constraints (meters)
 * 标高约束（米）
 */
export const ELEVATION_CONSTRAINTS = {
  min: -500, // 最低-500米（死海等特殊地区）
  max: 5000, // 最高5000米（高原机场）
};

File: backend/app/models/runway.py
"""Runway-related Pydantic models."""
from pydantic import BaseModel, Field, field_validator, ConfigDict
from typing import Literal, List


def to_camel(string: str) -> str:
    """Convert snake_case to camelCase."""
    components = string.split('_')
    return components[0] + ''.join(x.title() for x in components[1:])


class Coordinate(BaseModel):
    """Coordinate representation (坐标点表示)."""
    
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )
    
    latitude: float = Field(
        ...,
        description="纬度，单位：度",
        ge=-90,
        le=90,
        examples=[39.9]
    )
    longitude: float = Field(
        ...,
        description="经度，单位：度",
        ge=-180,
        le=180,
        examples=[116.4]
    )
    
    @field_validator('latitude', 'longitude')
    @classmethod
    def validate_coordinate_precision(cls, v: float) -> float:
        """Validate coordinate has reasonable precision."""
        # Round to 6 decimal places for ~0.1m precision
        return round(v, 6)


class ValidationError(BaseModel):
    """Validation error for runway parameters (跑道参数校验错误)."""
    
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )
    
    field: str = Field(..., description="错误字段名")
    message: str = Field(..., description="错误消息")
    severity: Literal["error", "warning"] = Field(
        default="error",
        description="错误级别"
    )


class RunwayValidationResult(BaseModel):
    """Runway validation result (跑道参数校验结果)."""
    
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )
    
    is_valid: bool = Field(..., description="是否有效")
    errors: List[ValidationError] = Field(
        default_factory=list,
        description="校验错误列表"
    )


class RunwayParams(BaseModel):
    """Runway parameters for flight circuit calculation (跑道参数定义)."""
    
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )
    
    coordinate: Coordinate = Field(
        ...,
        description="跑道中心点坐标"
    )
    magnetic_bearing: float = Field(
        ...,
        description="磁方位角，单位：度，范围0-360",
        ge=0,
        le=360,
        examples=[180]
    )
    length: int = Field(
        ...,
        description="跑道长度，单位：米",
        ge=200,
        le=5000,
        examples=[800]
    )
    elevation: float = Field(
        ...,
        description="跑道标高，单位：米",
        ge=-500,
        le=5000,
        examples=[50.0]
    )
    runway_width: float = Field(
        default=0,
        description="跑道宽度，单位：米（0=自动根据飞行区指标确定，>0=手动设定）",
        ge=0,
        le=100,
        examples=[18]
    )
    coordinate_system: Literal["WGS84", "GCJ02"] = Field(
        default="WGS84",
        description="输入坐标系类型"
    )
    
    @field_validator('magnetic_bearing')
    @classmethod
    def validate_magnetic_bearing(cls, v: float) -> float:
        """Normalize magnetic bearing to 0-360 range."""
        # Normalize to 0-360 if outside range
        if v < 0 or v > 360:
            v = v % 360
        return round(v, 1)
    
    @field_validator('length')
    @classmethod
    def validate_runway_length(cls, v: int) -> int:
        """Validate runway length is reasonable for light aircraft."""
        if v < 600:
            # Warning but still valid for ultra-light aircraft
            pass
        return v

    @field_validator('elevation')
    @classmethod
    def validate_elevation_precision(cls, v: float) -> float:
        """Keep runway elevation to 0.1m precision."""
        return round(v, 1)


class CoordinateInternal(BaseModel):
    """Internal coordinate with system tracking (内部坐标表示)."""
    
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )
    
    lat: float = Field(..., description="WGS84纬度")
    lng: float = Field(..., description="WGS84经度")
    original_system: Literal["WGS84", "GCJ02"] = Field(
        default="WGS84",
        description="用户输入时的坐标系"
    )
